### ISSUE TITLE
Update the Desktop Software

### ISSUE THREAD

**Issue by PythonLinks** [microCore-VHDL/microCore#4]
I think that the desktop software has awesome functionality. 
It can cross compile, generate the vhdl for the memory for the synthesizer, it can be used to debug. 

But perhaps it is a bit dated.  Did I hear you say it runs on a 6502 emulator???  Requires an RS232 cable, runs on an old version of gForth, which you had to do some really complex things to.  And requires docker.  I presume that it requires an Intel cpu.  I run on a M1 mac. 

it would be a most interesting project to build a Python emulator of your cpu, which could read your architecture file, act as a microForth interpreter, generate the VHDL for synthesis, and interactively debug the chip.   It would take a bunch of work to reproduce what you have, but once that was done, interesting things become possible.  

One could design a new Forth word, using some of the high level Python tools, and once it was debugged, export it to vhdl or Verilog for synthesis.   Also then I could use it from my M1 Mac.   There are a couple of other things I could imagine doing with it. 

Anyhow lots to learn here.    I do find this project fascinating.

**Comment by PythonLinks** [comment]
And here is what the updated client software would look like.  This is the Python software which emulates  James Bowman's J1a.  912 lines of code, should not be too difficult to make it emulate the microCore. 

https://github.com/jamesbowman/swapforth/blob/master/python/nuc.py

Really I want to  merge the best parts of both of your cpus.

**Comment by microCore-VHDL** [comment]
> Did I hear you say it runs on a 6502 emulator???

No. It runs on gforth 0.6.2., for which I made a docker image. As soon as gforth_1.0 finally becomes available, I will port the code to it without patching. This will then allow to generate code for instatiations up to a 64 bits.

> I presume that it requires an Intel cpu. I run on a M1 mac. 

See https://github.com/microCore-VHDL/microCore/blob/master/software/rs232_macosx.fs. This is the MAC RS232 interface. It has not been used for a long time though!

> it would be a most interesting project to build a Python emulator of your cpu, ...

I don't see your point. When I work with uCore, I always have a piece of HW with an FPGA. Then I port uCore to it, which usually means to write a new pin assignment file and a few timing constraints to get it going. Then I can do my tests on the "real thing" and use the debugger to gain insight. 
If you really want to do it: The most complex part is to compile a forward branch, which needs a number of literals in front of the branch instruction itself, which initially is not known. And the library mechanism is and probably will remain a real challenge.
Also, if the host is not a Forth system, it becomes very difficult to cross-compile new 'Create ... Does>' data types, because the code between Create and Does> has to be compiled on the host to be executed when the new data type is instantiated. 

> Also then I could use it from my M1 Mac. 

Is docker with a serial interface not available on the MAC?
I already found out that it can not be used on Win10, because Microsoft's docker does not support serial IO.
